portal/flatpak-portal.c

```diff
@@ -174,6 +176,8 @@ static guint idle_timeout_id = 0;
 static gboolean
 idle_timeout_cb (gpointer user_data)
 {
+  G_LOCK (client_pid_data_hash);
+
   if (name_owner_id &&
       g_hash_table_size (client_pid_data_hash) == 0 &&
       !has_update_monitors ())
```

Comment:
`has_update_monitors()` takes a different lock, which makes me a little concerned about lock ordering: if we try to take the `update_monitors` lock while already holding the `client_pid_data_hash` lock, and another thread is holding the `update_monitors` lock and trying to take the `client_pid_data_hash` lock, then we will deadlock.

Reply:
The other places that take this lock don't seem to do too much under the lock, though.

Reply:
I wonder whether `idle_timeout_cb()` would be better to be using something more like this:

````
if (!g_mutex_trylock (G_LOCK_NAME (client_pid_data_hash)))
  return G_SOURCE_CONTINUE;

if (!g_mutex_trylock (G_LOCK_NAME (update_monitors)))
  return G_SOURCE_CONTINUE;

if (name_owner_id &&
    g_hash_table_size (client_pid_data_hash) == 0 &&
    g_hash_table_size (update_monitors) == 0)
  {
      g_info ("Idle - unowning name");
      unref_skeleton_in_timeout ();
  }

  idle_timeout_id = 0;
  G_UNLOCK (update_monitors);
  G_UNLOCK (client_pid_data_hash);
  return G_SOURCE_REMOVE;
````

so that, if another thread happens to be currently doing things with the hash tables, it errs on the side of "I am not yet idle, try again later".